tests/resource_name_test.c

```diff
@@ -0,0 +1,128 @@
+        AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("ar:aws:cloudformation:us-east-1:1234567890:stack/FooBar");
+    struct aws_resource_name arn_06;
+    AWS_ZERO_STRUCT(arn_06);
+    ASSERT_FAILS(aws_resource_name_init_from_cur(&arn_06, &arn_string_06));
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(resource_name_tostring_test, s_test_resource_name_tostring)
+static int s_test_resource_name_tostring(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    struct aws_byte_buf buffer;
+    AWS_ZERO_STRUCT(buffer);
+    ASSERT_SUCCESS(aws_byte_buf_init(&buffer, allocator, 1600));
+
+    struct aws_resource_name arn_01 = {.partition = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("aws-us-gov"),
+                                       .service = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("iam"),
+                                       .region = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(""),
+                                       .account_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("123456789"),
+                                       .resource_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("group/crt")};
+    ASSERT_SUCCESS(aws_byte_buf_append_resource_name(&buffer, &arn_01));
+    ASSERT_TRUE(aws_byte_buf_eq_c_str(&buffer, "arn:aws-us-gov:iam::123456789:group/crt"));
+
+    aws_byte_buf_reset(&buffer, false);
+    struct aws_resource_name arn_02 = {.partition = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("aws"),
+                                       .service = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("cloudformation"),
+                                       .region = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("us-west-2"),
+                                       .account_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("12345678910"),
+                                       .resource_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("stack/MyStack")};
+    ASSERT_SUCCESS(aws_byte_buf_append_resource_name(&buffer, &arn_02));
+    ASSERT_TRUE(aws_byte_buf_eq_c_str(&buffer, "arn:aws:cloudformation:us-west-2:12345678910:stack/MyStack"));
+    aws_byte_buf_clean_up(&buffer);
+
+    struct aws_byte_buf too_small_buffer;
+    AWS_ZERO_STRUCT(too_small_buffer);
+    ASSERT_SUCCESS(aws_byte_buf_init(&too_small_buffer, allocator, 16));
+    struct aws_resource_name arn_03 = {.partition = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("aws-cn"),
+                                       .service = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("dynamodb"),
+                                       .region = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("cn-northwest-1"),
+                                       .account_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("123456789"),
+                                       .resource_id = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("Table/Books")};
+    ASSERT_SUCCESS(aws_byte_buf_append_resource_name(&too_small_buffer, &arn_03));
+    ASSERT_TRUE(aws_byte_buf_eq_c_str(&too_small_buffer, "arn:aws-cn:dynamodb:cn-northwest-1:123456789:Table/Books"));
```

Comment:
try to avoid ASSERT_TRUE as it makes test failure output less useful.
ASSERT_BIN_ARRAY_EQUALS and the like would be useful here.

Reply:
Changed all of the c_str_equals to ASSERT_BIN_ARRAY_EQUALS.